newt/project/project.go

```diff
@@ -198,177 +209,350 @@ func (proj *Project) Warnings() []string {
-		return nil
+	// If the repo isn't installed yet, this is an install, not an upgrade.
+	if op == INSTALL_OP_UPGRADE && curVer == nil {
+		op = INSTALL_OP_INSTALL
 	}
 
-	for _, newRepo := range repos {
-		curRepo, ok := proj.repos[newRepo.Name()]
-		if !ok {
-			proj.repos[newRepo.Name()] = newRepo
-			return proj.updateRepos(proj.SortedRepos())
+	var verb string
+	switch op {
+	case INSTALL_OP_INSTALL:
+		if !force {
+			verb = "install"
 		} else {
-			// Add any dependencies we might have found here.
-			for _, dep := range newRepo.Deps() {
-				newRepo.DownloadDesc()
-				newRepo.ReadDesc()
-				curRepo.AddDependency(dep)
-			}
+			verb = "reinstall"
 		}
+
+	case INSTALL_OP_UPGRADE:
+		verb = "upgrade"
+
+	case INSTALL_OP_SYNC:
+		verb = "sync"
+
+	default:
+		return "", util.FmtNewtError(
+			"internal error: invalid install op: %v", op)
 	}
 
-	return nil
+	msg := fmt.Sprintf("    %s %s ", verb, repoName)
+	if curVer != nil {
+		msg += fmt.Sprintf("(%s --> %s)", curVer.String(), destVer.String())
+	} else {
+		msg += fmt.Sprintf("(%s)", destVer.String())
+	}
+
+	return msg, nil
 }
 
-// Selects repositories from the global state that satisfy the specified
-// predicate.
-func (proj *Project) SelectRepos(pred func(r *repo.Repo) bool) []*repo.Repo {
-	all := proj.SortedRepos()
-	var filtered []*repo.Repo
+// Describes an imminent repo operation to the user.  In addition, prompts the
+// user for confirmation if the `-a` (ask) option was specified.
+func (proj *Project) installPrompt(repoList []*repo.Repo,
+	vm deprepo.VersionMap, op installOp, force bool, ask bool) (bool, error) {
 
-	for _, r := range all {
-		if pred(r) {
-			filtered = append(filtered, r)
+	if len(repoList) == 0 {
+		return true, nil
+	}
+
+	util.StatusMessage(util.VERBOSITY_DEFAULT,
+		"Making the following changes to the project:\n")
+
+	for _, r := range repoList {
+		curVer := proj.projState.GetInstalledVersion(r.Name())
+		destVer := vm[r.Name()]
+
+		msg, err := proj.installMessageOneRepo(
+			r.Name(), op, force, curVer, destVer)
+		if err != nil {
+			return false, err
 		}
+
+		util.StatusMessage(util.VERBOSITY_DEFAULT, "%s\n", msg)
 	}
+	util.StatusMessage(util.VERBOSITY_DEFAULT, "\n")
 
-	return filtered
-}
+	if !ask {
+		return true, nil
+	}
 
-func (proj *Project) updateRepos(repos []*repo.Repo) error {
-	for _, r := range repos {
-		if r.IsLocal() {
-			continue
+	for {
+		fmt.Printf("Proceed? [Y/n] ")
+		line, more, err := bufio.NewReader(os.Stdin).ReadLine()
+		if more || err != nil {
+			return false, util.ChildNewtError(err)
 		}
 
-		err := proj.checkDeps(r)
-		if err != nil {
-			return err
+		trimmed := strings.ToLower(strings.TrimSpace(string(line)))
+		if len(trimmed) == 0 || strings.HasPrefix(trimmed, "y") {
+			// User wants to proceed.
+			return true, nil
+		}
+
+		if strings.HasPrefix(trimmed, "n") {
+			// User wants to cancel.
+			return false, nil
 		}
+
+		// Invalid response.
+		fmt.Printf("Invalid response.  Please enter 'y' or 'n'.\n")
```

Comment:
A small nit! `Please enter 'y' or 'n'` since you already have a prompt with `[Y/n]`, and both `<ENTER>`, `YES`, `Yes` would work this extra message feels unnecessary (and not precise).

Reply:
I agree - the specifics in the message are unnecessary.  I will change it to:
```
Invalid response.
```